### PR TITLE
Edit existing sidebar item

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/24964

Resolves 
Screen Shot 2022-08-30 at 2 54 06 PM Screen Shot 2022-08-30 at 2 54 57 PM

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

`npm run test brave_unit_tests -- --filter=SidebarServiceTest*`

1. Launch Brave and add any two sites to sidebar
2. Check sidebar's builtin items(ex, bookmark) doesn't have `edit...` menu in context menu
3. Check newly added two items have `edit...` context menu
4. Try to modify url or title and check it's applied
5. When another item's url is used when modifying, it's not applied

### CODE REVIEW

File: browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      IDS_SIDEBAR_EDIT_ITEM_BUBBLE_AX_URL_EDITOR_LABLE));
+}
+
+views::View* SidebarEditItemBubbleDelegateView::GetInitiallyFocusedView() {
+  return title_tf_;
+}
+
+void SidebarEditItemBubbleDelegateView::UpdateItem() {
+  GetSidebarService(browser_)->UpdateItem(
+      target_item_.url, GURL(url_tf_->GetText()), title_tf_->GetText());
```
Comment (sangwoo108) [inline browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc:139]: I believe we can call this only when the data should changed.

File: browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+  title_tf_ = title_part->AddChildView(std::make_unique<views::Textfield>());
+  title_tf_->SetText(target_item_.title);
+  title_tf_->SelectAll(true);
+  title_tf_->SetAccessibleName(brave_l10n::GetLocalizedResourceUTF16String(
+      IDS_SIDEBAR_EDIT_ITEM_BUBBLE_AX_TITLE_EDITOR_LABLE));
+
+  auto* url_part = AddChildView(std::make_unique<views::View>());
+  url_part->SetLayoutManager(std::make_unique<views::BoxLayout>(
+      views::BoxLayout::Orientation::kVertical, gfx::Insets(), 4));
```
Comment (sangwoo108) [inline browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc:114]: nit:

```suggestion
      views::BoxLayout::Orientation::kVertical, gfx::Insets(), /* between_child_spacing = */ 4));
```

File: browser/ui/views/sidebar/sidebar_items_contents_view.cc [hunk 3]
```diff
@@ -174,6 +180,19 @@ void SidebarItemsContentsView::ShowContextMenuForViewImpl(
...
 }
 
+void SidebarItemsContentsView::LaunchEditItemDialog() {
+  DCHECK(!observation_.IsObserving());
+  DCHECK(view_for_context_menu_);
+  int index = GetIndexOf(view_for_context_menu_);
+  const auto& items = sidebar_model_->GetAllSidebarItems();
+
+  auto* bubble = views::BubbleDialogDelegateView::CreateBubble(
+      new SidebarEditItemBubbleDelegateView(browser_, items[index],
```
Comment (sangwoo108) [inline browser/ui/views/sidebar/sidebar_items_contents_view.cc:190]: I was surprised that `items[index]` is allowed here, as it's `const`

**Review by sangwoo108 — COMMENTED** [review]
Sorry for delayed review

**Comment by simonhong** [comment]
> Sorry for delayed review

NP! Thanks for review. I think I address all your comments. PTAL again :)

File: browser/ui/views/sidebar/sidebar_items_contents_view.cc (hunk 3 above)
Comment (sangwoo108) [reply browser/ui/views/sidebar/sidebar_items_contents_view.cc:190]: Ah, my bad. I misread items and thought it's a map. NVM.

File: components/sidebar/sidebar_service.h [hunk 4]
```diff
@@ -23,6 +23,14 @@ class PrefService;
 
 namespace sidebar {
 
+struct SidebarItemUpdate {
+  int index = -1;
+  bool title_updated = false;
+  bool url_updated = false;
+
+  bool operator==(const SidebarItemUpdate& update) const;
+};
```
Comment (sangwoo108) [inline components/sidebar/sidebar_service.h:32]: 👍

File: components/sidebar/sidebar_service.cc [hunk 5]
```diff
@@ -340,6 +345,34 @@ void SidebarService::MoveItem(int from, int to) {
...
+  }
+
+  auto item_iter = base::ranges::find_if(
+      items_, [&old_url](const auto& item) { return (item.url == old_url); });
+  if (item_iter != items_.end()) {
+    const int index = std::distance(items_.begin(), item_iter);
+    DCHECK(IsEditableItemAt(index));
+    item_iter->url = new_url;
+    item_iter->title = new_title;
+    for (Observer& obs : observers_)
```
Comment (sangwoo108) [inline components/sidebar/sidebar_service.cc:369]: nit : I believe wrapping for loop body with `{ }` complies with our convention, as the body is more than 1 physical line.

File: browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+void SidebarEditItemBubbleDelegateView::UpdateOKButtonEnabledState() {
+  DCHECK(title_tf_ && url_tf_);
+
+  // To update item, we need valid url.
+  bool ok_button_enabled = !url_tf_->GetText().empty();
+  if (ok_button_enabled && target_item_.url == GURL(url_tf_->GetText()) &&
+      target_item_.title == title_tf_->GetText()) {
+    ok_button_enabled = false;
+  }
```
Comment (sangwoo108) [inline browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc:165]: I think it's okay to pass empty check.

```suggestion
  // Update item only when url or title is changed
  const bool ok_button_enabled = target_item_.url != GURL(url_tf_->GetText()) ||
      target_item_.title != title_tf_->GetText()
```

File: browser/ui/views/sidebar/sidebar_add_item_bubble_delegate_view.cc [hunk 7]
```diff
@@ -170,6 +170,8 @@ void SidebarAddItemBubbleDelegateView::AddChildViews() {
         std::make_unique<SidebarAddItemButton>(true, color_provider));
     const GURL active_tab_url =
         browser_->tab_strip_model()->GetActiveWebContents()->GetVisibleURL();
+    LOG(ERROR) << __func__ << " active_tab_url :" << active_tab_url;
+    LOG(ERROR) << __func__ << " active_tab_url :" << active_tab_url.host();
```
Comment (sangwoo108) [inline browser/ui/views/sidebar/sidebar_add_item_bubble_delegate_view.cc:174]: oops :)

**Review by sangwoo108 — APPROVED** [review]
LGTM with nits

File: browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc (hunk 6 above)
Comment (simonhong) [reply browser/ui/views/sidebar/sidebar_edit_item_bubble_delegate_view.cc]: Changed to use original url when url input is empty.
If url is empty, we don't know which item's title should be updated when user set different title.

**Review by simonhong — COMMENTED** [review]
Done. thanks for review :)

**Comment by simonhong** [comment]
Merged after squshing.

**Pull request merged**